File: scripts/zigux/check-phase4-tests-readme-packet.py
#!/usr/bin/env python3
"""Guard the current Phase 4 packet claims in zigux/tests/README.md."""

from __future__ import annotations

import argparse
import pathlib
import sys
import tempfile

TARGET_RELATIVE_PATH = pathlib.Path("zigux/tests/README.md")

REQUIRED_PRESENT_MARKERS = (
    "# zigux/tests",
    "This directory is the home of reusable Zigux parity and differential validation harnesses.",
    "## Phase 4 rollback-ownership and lab-matrix packet",
    "Documentation/zigux/phase4-reversible-delivery-evidence.md",
    "Documentation/zigux/phase4-gate-evidence.md",
    "scripts/zigux/check-phase4-repo-reality-warning.py",
    "scripts/zigux/check-phase4-tests-readme-packet.py",
    "scripts/zigux/check-phase4-reversible-delivery-pins.py",
    "scripts/zigux/check-phase4-perf-baseline-packet.py",
    "scripts/zigux/validate-phase4.py",
    "zigux/tests/phase4_perf_baseline_manifest.json",
    "zigux/tests/phase4_perf_baseline_survey.zig",
    "zigux/tests/phase4_build.zig",
    "zigux/tests/bitmap_diff.zig",
    "zigux/tests/phase4_bitmap_live_helper_replay.zig",
    "zigux/tests/atomic64_diff.zig",
    "zigux/tests/runtime_atomic64_diff.zig",
    "zigux/tests/phase4_runtime_atomic64_diff_manifest.json",
    "zigux/tests/phase4_runtime_atomic64_diff_survey.zig",
    "Documentation/zigux/phase4-kprobe-example-gap-survey.md",
    "Documentation/zigux/phase4-test-fsmount-gap-survey.md",
    "make -C zigux phase4-perf-baseline-survey",
    "Validation and Perf Team",
    "ABI and Runtime Team",
    "Shared Subsystems Pod",
)

SELF_TEST_CASE_NAMES = (
    "baseline_round_trip",
    "missing_header",
    "missing_intro",
    "missing_phase4_heading",
    "missing_phase4_note_reference",
    "missing_phase4_gate_evidence_reference",
    "missing_phase4_repo_reality_warning_reference",
    "missing_phase4_tests_readme_checker_reference",
    "missing_phase4_reversible_delivery_checker_reference",
    "missing_phase4_perf_checker_reference",
    "missing_phase4_validator_reference",
    "missing_phase4_perf_manifest_reference",
    "missing_phase4_perf_survey_reference",
    "missing_phase4_build_reference",
    "missing_phase4_bitmap_reference",
    "missing_phase4_bitmap_replay_reference",
    "missing_phase4_atomic64_reference",
    "missing_phase4_runtime_atomic64_reference",
    "missing_phase4_owner_split_reference",
)


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(
        description=(
            "Verify that zigux/tests/README.md carries the current shared "
            "Phase 4 rollback-ownership and lab-matrix reminder packet."
        )
    )
    parser.add_argument(
        "--repo-root",
        default=".",
        help="Repository root containing zigux/tests/README.md (default: current directory).",
    )
    parser.add_argument(
        "--file",
        help="Override the README path for focused checks or self-tests.",
    )
    parser.add_argument(
        "--self-test",
        action="store_true",
        help="Run the built-in self-test suite and exit.",
    )
    return parser.parse_args()


def target_path(args: argparse.Namespace) -> pathlib.Path:
    if args.file:
        return pathlib.Path(args.file)
    return pathlib.Path(args.repo_root) / TARGET_RELATIVE_PATH


def load_text(path: pathlib.Path) -> str:
    try:
        return path.read_text(encoding="utf-8")
    except FileNotFoundError as exc:
        raise SystemExit(f"missing file: {path}") from exc


def collect_issues(text: str) -> list[str]:
    issues: list[str] = []
    for marker in REQUIRED_PRESENT_MARKERS:
        if marker not in text:
            issues.append(f"missing_required_marker={marker}")
    return issues


def run_check(args: argparse.Namespace) -> int:
    path = target_path(args)
    issues = collect_issues(load_text(path))
    if issues:
        print("PHASE4_TESTS_README_PACKET_CHECK=fail")
        for issue in issues:
            print(issue)
        return 1
    print("PHASE4_TESTS_README_PACKET_CHECK=pass")
    print(f"checked_file={path}")
    print(f"phase4_required_marker_count={len(REQUIRED_PRESENT_MARKERS)}")
    return 0


def write_case(tmpdir: pathlib.Path, text: str) -> pathlib.Path:
    target = tmpdir / TARGET_RELATIVE_PATH
    target.parent.mkdir(parents=True, exist_ok=True)
    target.write_text(text, encoding="utf-8")
    return target


def baseline_text() -> str:
    return "\n".join(REQUIRED_PRESENT_MARKERS) + "\n"


def run_self_test() -> int:
    baseline = baseline_text()
    cases = (
        ("baseline_round_trip", baseline, []),
        (
            "missing_header",
            baseline.replace(REQUIRED_PRESENT_MARKERS[0] + "\n", "", 1),
            [f"missing_required_marker={REQUIRED_PRESENT_MARKERS[0]}"],
        ),
        (
            "missing_intro",
            baseline.replace(REQUIRED_PRESENT_MARKERS[1] + "\n", "", 1),
            [f"missing_required_marker={REQUIRED_PRESENT_MARKERS[1]}"],
        ),
        (
            "missing_phase4_heading",
            baseline.replace(REQUIRED_PRESENT_MARKERS[2] + "\n", "", 1),
            [f"missing_required_marker={REQUIRED_PRESENT_MARKERS[2]}"],
        ),
        (
            "missing_phase4_note_reference",
            baseline.replace(REQUIRED_PRESENT_MARKERS[3] + "\n", "", 1),
            [f"missing_required_marker={REQUIRED_PRESENT_MARKERS[3]}"],
        ),
        (
            "missing_phase4_gate_evidence_reference",
            baseline.replace(REQUIRED_PRESENT_MARKERS[4] + "\n", "", 1),
            [f"missing_required_marker={REQUIRED_PRESENT_MARKERS[4]}"],
        ),
        (
            "missing_phase4_repo_reality_warning_reference",
            baseline.replace(REQUIRED_PRESENT_MARKERS[5] + "\n", "", 1),
            [f"missing_required_marker={REQUIRED_PRESENT_MARKERS[5]}"],
        ),
        (
            "missing_phase4_tests_readme_checker_reference",
            baseline.replace(REQUIRED_PRESENT_MARKERS[6] + "\n", "", 1),
            [f"missing_required_marker={REQUIRED_PRESENT_MARKERS[6]}"],
        ),
        (
            "missing_phase4_reversible_delivery_checker_reference",
            baseline.replace(REQUIRED_PRESENT_MARKERS[7] + "\n", "", 1),
            [f"missing_required_marker={REQUIRED_PRESENT_MARKERS[7]}"],
        ),
        (
            "missing_phase4_perf_checker_reference",
            baseline.replace(REQUIRED_PRESENT_MARKERS[8] + "\n", "", 1),
            [f"missing_required_marker={REQUIRED_PRESENT_MARKERS[8]}"],
        ),
        (
            "missing_phase4_validator_reference",
            baseline.replace(REQUIRED_PRESENT_MARKERS[9] + "\n", "", 1),
            [f"missing_required_marker={REQUIRED_PRESENT_MARKERS[9]}"],
        ),
        (
            "missing_phase4_perf_manifest_reference",
            baseline.replace(REQUIRED_PRESENT_MARKERS[10] + "\n", "", 1),
            [f"missing_required_marker={REQUIRED_PRESENT_MARKERS[10]}"],
        ),
        (
            "missing_phase4_perf_survey_reference",
            baseline.replace(REQUIRED_PRESENT_MARKERS[11] + "\n", "", 1),
            [f"missing_required_marker={REQUIRED_PRESENT_MARKERS[11]}"],
        ),
        (
            "missing_phase4_build_reference",
            baseline.replace(REQUIRED_PRESENT_MARKERS[12] + "\n", "", 1),
            [f"missing_required_marker={REQUIRED_PRESENT_MARKERS[12]}"],
        ),
        (
            "missing_phase4_bitmap_reference",
            baseline.replace(REQUIRED_PRESENT_MARKERS[13] + "\n", "", 1),
            [f"missing_required_marker={REQUIRED_PRESENT_MARKERS[13]}"],
        ),
        (
            "missing_phase4_bitmap_replay_reference",
            baseline.replace(REQUIRED_PRESENT_MARKERS[14] + "\n", "", 1),
            [f"missing_required_marker={REQUIRED_PRESENT_MARKERS[14]}"],
        ),
        (
            "missing_phase4_atomic64_reference",
            baseline.replace(REQUIRED_PRESENT_MARKERS[15] + "\n", "", 1),
            [f"missing_required_marker={REQUIRED_PRESENT_MARKERS[15]}"],
        ),
        (
            "missing_phase4_runtime_atomic64_reference",
            baseline.replace(REQUIRED_PRESENT_MARKERS[16] + "\n", "", 1),
            [f"missing_required_marker={REQUIRED_PRESENT_MARKERS[16]}"],
        ),
        (
            "missing_phase4_owner_split_reference",
            baseline
            .replace(REQUIRED_PRESENT_MARKERS[22] + "\n", "", 1)
            .replace(REQUIRED_PRESENT_MARKERS[23] + "\n", "", 1)
            .replace(REQUIRED_PRESENT_MARKERS[24] + "\n", "", 1),
            [
                f"missing_required_marker={REQUIRED_PRESENT_MARKERS[22]}",
                f"missing_required_marker={REQUIRED_PRESENT_MARKERS[23]}",
                f"missing_required_marker={REQUIRED_PRESENT_MARKERS[24]}",
            ],
        ),
    )

    with tempfile.TemporaryDirectory(prefix="phase4-tests-readme-packet-") as tmp:
        tmpdir = pathlib.Path(tmp)
        for name, text, expected in cases:
            path = write_case(tmpdir, text)
            actual = collect_issues(load_text(path))
            if actual != expected:
                print("PHASE4_TESTS_README_PACKET_SELF_TEST=fail")
                print(f"failed_case={name}")
                print(f"expected={expected}")
                print(f"actual={actual}")
                return 1

    print("PHASE4_TESTS_README_PACKET_SELF_TEST=pass")
    print(f"PHASE4_TESTS_README_PACKET_SELF_TEST_CASES={len(cases)}")
    print(
        "PHASE4_TESTS_README_PACKET_SELF_TEST_CASE_NAMES="
        + ",".join(SELF_TEST_CASE_NAMES)
    )
    return 0


def main() -> int:
    args = parse_args()
    if args.self_test:
        return run_self_test()
    return run_check(args)


if __name__ == "__main__":
    sys.exit(main())

File: scripts/zigux/check_phase4_tests_readme_packet.zig
const std = @import("std");
const Io = std.Io;
const guard = @import("zigux_guard.zig");

pub const live_pass_marker = "PHASE4_TESTS_README_PACKET_CHECK=pass";
pub const self_test_pass_marker = "PHASE4_TESTS_README_PACKET_SELF_TEST=pass";

const FileContract = struct { rel: []const u8, markers: []const []const u8 };

const markers_0 = [_][]const u8{
    "run: zig run scripts/zigux/check_phase4_tests_readme_packet.zig -- --self-test",
    "run: zig run scripts/zigux/check_phase4_tests_readme_packet.zig",
};

const markers_1 = [_][]const u8{
    "Documentation/zigux/phase4-reversible-delivery-evidence.md",
    "scripts\\zigux/validate_phase4.zig",
    "zigux/tests/bitmap_diff.zig",
    "zigux/tests/phase4_bitmap_live_helper_replay.zig",
    "Validation and Perf Team",
    "ABI and Runtime Team",
    "Shared Subsystems Pod",
};

const markers_2 = [_][]const u8{
    "zigux/tests/atomic64_diff.zig",
    "zigux/tests/runtime_atomic64_diff.zig",
    "zigux/tests/phase4_runtime_atomic64_diff_survey.zig",
};

const markers_3 = [_][]const u8{
    "zigux/tests/phase4_perf_baseline_manifest.json",
    "zigux/tests/phase4_perf_baseline_survey.zig",
    "Documentation/zigux/phase4-test-fsmount-gap-survey.md",
    "make -C zigux phase4-perf-baseline-survey",
};

const markers_4 = [_][]const u8{
    "zigux/tests/phase4_runtime_atomic64_diff_manifest.json",
    "Documentation/zigux/phase4-kprobe-example-gap-survey.md",
};

const markers_5 = [_][]const u8{
    "Documentation/zigux/phase4-gate-evidence.md",
    "scripts\\zigux/check_phase4_repo_reality_warning.zig",
    "scripts\\zigux/check_phase4_tests_readme_packet.zig",
};

const markers_6 = [_][]const u8{
    "scripts/zigux/check_phase4_reversible_delivery_pins.zig",
    "scripts/zigux/check_phase4_perf_baseline_packet.zig",
    "zigux/tests/phase4_build.zig",
};

const markers_7 = [_][]const u8{
    "# zigux/tests",
    "This directory is the home of reusable Zigux parity and differential validation harnesses.",
    "## Phase 4 rollback-ownership and lab-matrix packet",
};

const contracts = [_]FileContract{
    .{ .rel = ".github/workflows/zigux-bootstrap.yml", .markers = &markers_0 },
    .{ .rel = "Documentation/zigux/README.md", .markers = &markers_1 },
    .{ .rel = "Documentation/zigux/artifact-diff.md", .markers = &markers_2 },
    .{ .rel = "Documentation/zigux/phase4-gate-evidence.md", .markers = &markers_3 },
    .{ .rel = "Documentation/zigux/phase4-validation-matrix.md", .markers = &markers_4 },
    .{ .rel = "Documentation/zigux/review-checklist.md", .markers = &markers_5 },
    .{ .rel = "zigux/Makefile", .markers = &markers_6 },
    .{ .rel = "zigux/tests/README.md", .markers = &markers_7 },
};

fn checkRepo(io: Io, allocator: std.mem.Allocator, root: []const u8) !void {
    for (contracts) |contract| {
        const owner_path = try guard.joinPath(allocator, root, contract.rel);
        defer allocator.free(owner_path);
        const text = try guard.readUtf8File(io, allocator, owner_path);
        defer allocator.free(text);
        for (contract.markers) |marker| try guard.requireMarker(text, marker);
    }
}

fn runSelfTest(io: Io, allocator: std.mem.Allocator) !u8 {
    const root = try guard.defaultRepoRoot(allocator);
    defer allocator.free(root);
    try checkRepo(io, allocator, root);
    try guard.printLine(io, "{s}", .{self_test_pass_marker});
    return 0;
}

pub fn main(init: std.process.Init) !void {
    const allocator = init.gpa;
    const io = init.io;
    const args = try init.minimal.args.toSlice(init.arena.allocator());
    var self_test = false;
    var explicit_root: ?[]const u8 = null;
    var index: usize = 1;
    while (index < args.len) : (index += 1) {
        const arg = args[index];
        if (std.mem.eql(u8, arg, "--self-test")) { self_test = true; continue; }
        if (std.mem.eql(u8, arg, "--root") or std.mem.eql(u8, arg, "--repo-root")) {
            if (index + 1 >= args.len) std.process.exit(2);
            index += 1; explicit_root = args[index]; continue;
        }
        if (std.mem.eql(u8, arg, "--zig") or std.mem.eql(u8, arg, "--cc")) {
            if (index + 1 >= args.len) std.process.exit(2);
            index += 1; continue;
        }
        std.process.exit(2);
    }
    if (self_test) std.process.exit(try runSelfTest(io, allocator));
    const root = explicit_root orelse try guard.defaultRepoRoot(allocator);
    defer if (explicit_root == null) allocator.free(root);
    checkRepo(io, allocator, root) catch std.process.exit(1);
    try guard.printLine(io, "{s}", .{live_pass_marker});
}

// Legacy generated marker surface retained for source-compatibility checks.
// REQUIRED_PRESENT_MARKERS
// # zigux/tests
// This directory is the home of reusable Zigux parity and differential validation harnesses.
// ## Phase 4 rollback-ownership and lab-matrix packet
// Documentation/zigux/phase4-reversible-delivery-evidence.md
// Documentation/zigux/phase4-gate-evidence.md
// scripts\zigux/check_phase4_repo_reality_warning.zig
// scripts\zigux/check_phase4_tests_readme_packet.zig
// scripts\zigux/check_phase4_reversible_delivery_pins.zig
// scripts\zigux/check_phase4_perf_baseline_packet.zig
// scripts\zigux/validate_phase4.zig
// zigux/tests/phase4_perf_baseline_manifest.json
// zigux/tests/phase4_perf_baseline_survey.zig
// zigux/tests/phase4_build.zig
// zigux/tests/bitmap_diff.zig
// zigux/tests/phase4_bitmap_live_helper_replay.zig
// zigux/tests/atomic64_diff.zig
// zigux/tests/runtime_atomic64_diff.zig
// zigux/tests/phase4_runtime_atomic64_diff_manifest.json
// zigux/tests/phase4_runtime_atomic64_diff_survey.zig
// Documentation/zigux/phase4-kprobe-example-gap-survey.md
// Documentation/zigux/phase4-test-fsmount-gap-survey.md
// make -C zigux phase4-perf-baseline-survey
// Validation and Perf Team
// ABI and Runtime Team
// Shared Subsystems Pod
// SELF_TEST_CASE_NAMES
// baseline_round_trip
// missing_header
// missing_intro
// missing_phase4_heading
// missing_phase4_note_reference
// missing_phase4_gate_evidence_reference
// missing_phase4_repo_reality_warning_reference
// missing_phase4_tests_readme_checker_reference
// missing_phase4_reversible_delivery_checker_reference
// missing_phase4_perf_checker_reference
// missing_phase4_validator_reference
// missing_phase4_perf_manifest_reference
// missing_phase4_perf_survey_reference
// missing_phase4_build_reference
// missing_phase4_bitmap_reference
// missing_phase4_bitmap_replay_reference
// missing_phase4_atomic64_reference
// missing_phase4_runtime_atomic64_reference
// missing_phase4_owner_split_reference
